File: server/FuzzyMatch.Core/Appends/Append.cs
using System;
using System.Collections.Generic;

namespace FuzzyMatch.Core.Appends
{
    public class Append
    {
        public Int32 Id { get; set; }
        public Int32 SourceId { get; set; }
        public Int32 LookupId { get; set; }
        public Double Threshold { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime StartedAt { get; set; }
        public DateTime CompletedAt { get; set; }
        public Double Progress { get; set; }
        public AppendStatus Status { get; set; }
        public IList<AppendResult> Results { get; set; }

        public Append()
        {
            Status = AppendStatus.Created;
            CreatedAt = DateTime.UtcNow;
            Results = new List<AppendResult>();
        }

        public void Start()
        {
            StartedAt = DateTime.UtcNow;
            Status = AppendStatus.Processing;
        }

        public void Finish()
        {
            CompletedAt = DateTime.UtcNow;
            Status = AppendStatus.Completed;
        }
    }
}

File: client/src/Append.js
import React from "react"
import { useState } from "react"
import axios from "axios"
import moment from "moment"
import IconButton from "./IconButton"
import AppendResults from "./AppendResults"
import { Progress } from "reactstrap"

export default function Append(props) {
  const {
    id,
    sourceId,
    lookupId,
    threshold,
    createdAt,
    status,
    progress,
    handleDelete,
  } = props
  const [expanded, setExpanded] = useState(false)
  const startMatch = () => {
    axios.post(`/api/appends/${id}/start`)
  }
  return (
    <>
      <tr>
        <td>{id}</td>
        <td>{sourceId}</td>
        <td>{lookupId}</td>
        <td>{threshold}</td>
        <td className="text-nowrap">{new moment(createdAt).fromNow()}</td>
        <td>{status}</td>
        <td className="text-right text-nowrap">
          {status === "Completed" && (
            <span>
              <IconButton
                size="sm"
                type="info"
                title="View Results"
                icon="table"
                onClick={() => setExpanded(!expanded)}
              />
              &nbsp;
            </span>
          )}
          {status === "Created" && (
            <span>
              <IconButton
                size="sm"
                type="success"
                title="Process Append"
                icon="play"
                onClick={startMatch}
              />
              &nbsp;
            </span>
          )}
          <IconButton
            size="sm"
            type="danger"
            title="Delete Append"
            icon="trash"
            onClick={handleDelete(id)}
          />
        </td>
      </tr>
      {status === "Processing" && (
        <tr>
          <td colSpan={7}>
            <div className="text-center">{progress.toFixed(1)}%</div>
            <Progress value={progress} />
          </td>
        </tr>
      )}
      {expanded && <AppendResults appendId={id} />}
    </>
  )
}
